presto-main/src/main/java/com/facebook/presto/tracing/testing/TestingTracer.java

```diff
@@ -0,0 +1,96 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.tracing.testing;
+
+import com.facebook.presto.spi.tracing.Tracer;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class TestingTracer
+        implements Tracer
+{
+    public final Map<String, TestingBlock> blockMap = new ConcurrentHashMap<>();
+    public final Map<String, TestingBlock> recorderBlockMap = new LinkedHashMap<>();
+    public final List<TestingPoint> pointList = new ArrayList<>();
+
+    public TestingTracer()
+    {
+        addPoint("Start tracing");
+    }
+
+    @Override
+    public synchronized void addPoint(String annotation)
+    {
+        pointList.add(new TestingPoint(annotation));
+    }
+
+    @Override
+    public void startBlock(String blockName, String annotation)
+    {
+        if (blockMap.containsKey(blockName)) {
+            throw new RuntimeException("Duplicated block inserted: " + blockName);
+        }
+        TestingBlock block = new TestingBlock(annotation);
+        blockMap.put(blockName, block);
+        synchronized (recorderBlockMap) {
+            recorderBlockMap.put(blockName, block);
+        }
+    }
+
+    @Override
+    public void addPointToBlock(String blockName, String annotation)
+    {
+        if (!blockMap.containsKey(blockName)) {
+            throw new RuntimeException("Adding point to non-existing block: " + blockName);
+        }
+        blockMap.get(blockName).addPoint(new TestingPoint(annotation));
+    }
+
+    @Override
+    public void endBlock(String blockName, String annotation)
+    {
+        if (!blockMap.containsKey(blockName)) {
+            throw new RuntimeException("Trying to end a non-existing block: " + blockName);
+        }
+        blockMap.remove(blockName);
+        synchronized (recorderBlockMap) {
+            recorderBlockMap.get(blockName).end(annotation);
+        }
+    }
+
+    @Override
+    public void endTrace(String annotation)
+    {
+        synchronized (pointList) {
+            pointList.add(new TestingPoint(annotation));
+        }
+        // Uncomment to speculate tracing details after each trace.
+        // System.out.println(getTracingDetails().toString());
+    }
+
+    private String getTracingDetails()
+    {
+        return new StringBuilder()
+                .append("{\n")
+                .append("\"blocks\":")
+                .append(recorderBlockMap.toString().replace('=', ':')).append(",")
+                .append("\"points\":")
+                .append(pointList.toString())
+                .append("}").toString();
+    }
```

Comment:
remove if this PR is ready; the general rule of thumb is to have unit tests to do "assertion" instead of "debugging". So probably we need more assertion checks to capture the logs we want to print out

Reply:
There are 2 things we want to test:
1. We want to test if TestingTracer is written correctly. This is done by TestTestingTracer(will be renamed as mentioned in the comments)
2. We want to test if tracing instrumentations in the codebase are put in the correct places. The TestingTracer should be enabled in some existing integration tests to test the instrumentations.